### ISSUE TITLE
How to train multi-scale VAE of VAR by XQGAN？

### ISSUE THREAD

**Issue by Pride-Huang** [lxa9867/ImageFolder#11]
I want to train a multi-scale VAE. Can I achieve it by XQGAN?

**Comment by qiuk2** [comment]
Thanks for your interest in our work! Our tokenizer supports multi-scale VQVAE. The training cmd is 

```
torchrun --nproc_per_node=8 tokenizer/tokenizer_image/xqgan_train.py --config configs/MSVR10P2-4096.yaml
```
If you need to train a multi-scale VAE from our framework, I think you can follow VAE to change our models/quant.py.

**Comment by Pride-Huang** [comment]
Thanks for your reply. And I find product_quant=2, which decomposes the embedding into two part. However, when I set it to 1, there is a bug in xqgan_model.py line 307, for quant_list not defined.

**Comment by qiuk2** [comment]
Hi, thanks for your patience, we currently unsupport pq=1 where simultaneously use dino as a distilling method. If you wish to continue working on pq=1 and the dino distilling method, you can define 
```
quant_list = [quant]
```
after line 292. Thanks for your help in validating our code. If there is more problems, please let me know.

**Comment by Pride-Huang** [comment]
Thanks. I have a another question, there is ten codebook usage. What does it mean? And how can I improve the usage to 95%?

![Image](https://github.com/user-attachments/assets/f52aec94-b0bb-49b3-ac04-1a27e165ebfe)

**Comment by qiuk2** [comment]
the codebook usage is shown in codebook_usage. It is a list representing the codebook usage in each layer (residual). In my experience, the codebook usage shown at first is similar to the figure you showed. If you can continue training, the codebook usage will increase and finally, the last layer's codebook usage will be 100

**Comment by Pride-Huang** [comment]
Many Thanks. The train loss now is 0.3841. How many epoch  is needed and what is a suitable value of train loss?

**Comment by qiuk2** [comment]
Currently, I do not know your training setting, but normally we will train 200 epochs in ImageNet and the total training loss will be around 0.5xxx. We use rFID to save our best checkpoints, which is around 0.8.

**Comment by Pride-Huang** [comment]
Pretty interesting！When I continue trainging, the loss have increased from 0.3841 to 0.8480, and the usage has been 95%, and FID is 0.89.  Anything else I need to do? Or just keep training?

![Image](https://github.com/user-attachments/assets/09b64813-6dbe-49b4-943d-62529dbe0085)

**Comment by qiuk2** [comment]
just keep training, it is normal

**Comment by qiuk2** [comment]
Since there are no more other questions, I will close this issue.

**Comment by Pride-Huang** [comment]
> the codebook usage is shown in codebook_usage. It is a list representing the codebook usage in each layer (residual). In my experience, the codebook usage shown at first is similar to the figure you showed. If you can continue training, the codebook usage will increase and finally, the last layer's codebook usage will be 100

Thanks. My problem has been solved. If it is possible, I also want to know why the codebook usage will be decreased, and then increased.

**Comment by qiuk2** [comment]
We cannot fully explain it but I think it may be cause we train our tokenizer from DINOv2 weight, and it shifts from latent space for understanding to reconstruction.
